tests/integration/test_nginx_route.py

```diff
@@ -0,0 +1,192 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+# pylint: disable=redefined-outer-name,unused-argument,duplicate-code
+
+"""Integration test relation file."""
+
+import asyncio
+import json
+import time
+import typing
+from pathlib import Path
+
+import pytest
+import pytest_asyncio
+import requests
+from juju.model import Model
+from pytest_operator.plugin import OpsTest
+
+INGRESS_APP_NAME = "ingress"
+ANY_APP_NAME = "any"
+
+
+def gen_src_overwrite(
+    service_hostname: str = "any",
+    service_name: str = "any",
+    service_port: int = 8080,
+    additional_hostnames: typing.Optional[str] = None,
+) -> str:
+    """Generate the src-overwrite config value for testing nginx-route relation."""
+    nginx_route_lib_path = "lib/charms/nginx_ingress_integrator/v0/nginx_route.py"
+    nginx_route_lib = Path(nginx_route_lib_path).read_text(encoding="utf8")
+    any_charm_script = Path("tests/integration/any_charm_nginx_route.py").read_text(
+        encoding="utf8"
+    )
+    nginx_route_config = {
+        "service_hostname": service_hostname,
+        "service_name": service_name,
+        "service_port": service_port,
+    }
+    if additional_hostnames:
+        nginx_route_config["additional_hostnames"] = additional_hostnames
+    any_charm_src_overwrite = {
+        "any_charm.py": any_charm_script,
+        "nginx_route.py": nginx_route_lib,
+        "nginx_route_config.json": json.dumps(nginx_route_config),
+    }
+    return json.dumps(any_charm_src_overwrite)
+
+
+def requests_get(url: str, host_header: str, retry_timeout: int = 120) -> requests.Response:
+    """Requests get, but will retry when the response status code is not 200."""
+    time_start = time.time()
+    while True:
+        response = requests.get(url, headers={"Host": host_header}, timeout=5)
+        if response.status_code == 200 or time.time() - time_start > retry_timeout:
+            return response
+
+
+@pytest_asyncio.fixture(scope="module")
+async def build_and_deploy(ops_test: OpsTest, run_action):
+    """build and deploy nginx-ingress-integrator charm.
+
+    Also deploy and relate an any-charm application for test purposes.
+
+    Returns: None.
+    """
+
+    assert isinstance(ops_test.model, Model)
```

Comment:
What about putting this into a fixture so this assert doesn't have to be done all over the place?